src/openms/source/ANALYSIS/XLMS/OpenPepXLAlgorithm.cpp

```diff
@@ -465,10 +479,134 @@ using namespace OpenMS;
+          }
+          OPXLSpectrumProcessingAlgorithms::getSpectrumAlignmentSimple(matched_spec_xlinks_alpha, fragment_mass_tolerance_xlinks_, fragment_mass_tolerance_unit_ppm_, theoretical_spec_xlinks_alpha, xlink_peaks, exp_charges);
+          OPXLSpectrumProcessingAlgorithms::getSpectrumAlignmentSimple(matched_spec_xlinks_beta, fragment_mass_tolerance_xlinks_, fragment_mass_tolerance_unit_ppm_, theoretical_spec_xlinks_beta, xlink_peaks, exp_charges);
+        }
+        // maximal xlink ion charge = (Precursor charge - 1), minimal xlink ion charge: 2
+        Size n_xlink_charges = (precursor_charge - 1) - 2;
+        if (n_xlink_charges < 1) n_xlink_charges = 1;
+
+        // compute match odds (unweighted), the 3 is the number of charge states in the theoretical spectra
+        double match_odds_c_alpha = XQuestScores::matchOddsScoreSimpleSpec(theoretical_spec_linear_alpha, matched_spec_linear_alpha.size(), fragment_mass_tolerance_, fragment_mass_tolerance_unit_ppm_);
+        double match_odds_x_alpha = XQuestScores::matchOddsScoreSimpleSpec(theoretical_spec_xlinks_alpha, matched_spec_xlinks_alpha.size(), fragment_mass_tolerance_xlinks_, fragment_mass_tolerance_unit_ppm_, true, n_xlink_charges);
+        double match_odds = 0;
+        double match_odds_alpha = 0;
+        double match_odds_beta = 0;
+
+        if (type_is_cross_link)
+        {
+          double match_odds_c_beta = XQuestScores::matchOddsScoreSimpleSpec(theoretical_spec_linear_beta, matched_spec_linear_beta.size(), fragment_mass_tolerance_, fragment_mass_tolerance_unit_ppm_);
+          double match_odds_x_beta = XQuestScores::matchOddsScoreSimpleSpec(theoretical_spec_xlinks_beta, matched_spec_xlinks_beta.size(), fragment_mass_tolerance_xlinks_, fragment_mass_tolerance_unit_ppm_, true, n_xlink_charges);
+          match_odds = (match_odds_c_alpha + match_odds_x_alpha + match_odds_c_beta + match_odds_x_beta) / 4;
+          match_odds_alpha = (match_odds_c_alpha + match_odds_x_alpha) / 2;
+          match_odds_beta = (match_odds_c_beta + match_odds_x_beta) / 2;
+        }
+        else
+        {
+          match_odds = (match_odds_c_alpha + match_odds_x_alpha) / 2;
+          match_odds_alpha = match_odds;
+        }
+
+        OPXLDataStructs::CrossLinkSpectrumMatch csm;
+        csm.cross_link = cross_link_candidate;
+        csm.precursor_correction = cross_link_candidate.precursor_correction;
+        double rel_error = OPXLHelper::computePrecursorError(csm, precursor_mz, precursor_charge);
+
+        double new_match_odds_weight = 0.2;
+        double new_rel_error_weight = -0.03;
+        double new_score = new_match_odds_weight * std::log(1e-7 + match_odds) + new_rel_error_weight * abs(rel_error);
+
+        csm.score = new_score;
+        csm.match_odds = match_odds;
+        csm.match_odds_alpha = match_odds_alpha;
+        csm.match_odds_beta = match_odds_beta;
+        csm.precursor_error_ppm = rel_error;
+
+        mainscore_csms_spectrum.push_back(csm);
```

Comment:
is it thread save to add it here? I did not check the whole context...

Reply:
This vector only exists within the context of each thread. It holds the main score of each candidate for one spectrum.